### PR TITLE
octeon: support the UniFi Security Gateway 3P

### PR DESCRIPTION
**Pull request by MartB**
Added support for the Ubiquiti Unifi Security Gateway

The default_packages edit is due to a failure to create the overlay filesystem because of a missing mkfs.f2fs binary.

The devicetree related edits are mandatory as theres no other established way (known to me) to load the proper devicetree on bootup on these devices and the USG3P has 2 leds (blue, white) i wanted to support.

If theres another way to do the device tree related changes and achieve the same result please tell me!

Compile and run tested on an USG3P.
Would be great if someone could test this on an edge router lite.

### CODE REVIEW

File: target/linux/octeon/Makefile [hunk 1]
```diff
@@ -21,6 +21,6 @@ endef
 
 include $(INCLUDE_DIR)/target.mk
 
-DEFAULT_PACKAGES += mkf2fs e2fsprogs
+DEFAULT_PACKAGES += mkf2fs e2fsprogs f2fs-tools
```
Comment (stintel) [inline target/linux/octeon/Makefile]: This is unrelated and should go in a different commit, if it should go in at all.

Comment (MartB) [reply target/linux/octeon/Makefile]: Tbh im not sure if it should go in but its the only way i got this working because the overlay filesystem kept failing.

Comment (stintel) [reply target/linux/octeon/Makefile]: That's strange. I'm using the squashfs + f2fs image on my ERLite and it works fine. I assume you are testing on master, as the PR is against master? Can you give more info on how it fails?

Comment (MartB) [reply target/linux/octeon/Makefile]: When i leave that package out it says "overlay filesystem has not been fully initialized yet." and then it switches to a temporary one that does hold the data until next reboot. I traced it down to the missing mkfs binary for f2fs. As soon as it was available it worked.

Comment (stintel) [reply target/linux/octeon/Makefile]: But f2fs-tools doesn't contain the mkfs utility:
```
root@ar1:~# opkg files f2fs-tools
Package f2fs-tools (1.8.0-3) is installed on root and has the following files:
/usr/sbin/f2fstat
/usr/sbin/parse.f2fs
/usr/sbin/fibmap.f2fs
```

Comment (stintel) [reply target/linux/octeon/Makefile]: Maybe you just didn't have mkf2fs enabled in your config initially? Having something in DEFAULT_PACKAGES does not guarantee it will be enabled in the .config. I suggest to disable f2fs-tools in menuconfig again and try a new build. After seeing what files are in there, it definitely does not belong in DEFAULT_PACKAGES.

I'll gladly test on my ERL when you update the PR.

Comment (MartB) [reply target/linux/octeon/Makefile]: I will do everything you mentioned above and update this pr.
I will also check if the mkfs.f2fs gets compiled and is available to the system.
I verified it got enabled in my configs.

**Comment by MartB** [comment]
Do not review at the moment im using this git repo on my build system and didnt branch out sorry!
I will remove the WIP as soon as it compiles.

**Comment by MartB** [comment]
I split the patches on 60bc430 into two separate files to get the smallest diff possible that does not contain the whole source from the linux kernel i hope its okay like this.

**Comment by stintel** [comment]
Not yet. I would like to see the patches exactly as how they were sent upstream by the original author, including the headers. E.g. https://github.com/lede-project/source/blob/master/target/linux/generic/backport-4.9/010-Kbuild-don-t-hardcode-path-to-awk-in-scripts-ld-vers.patch

File: target/linux/octeon/patches-4.9/103-ubnt_usg3p_support.patch [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
++extern const char __dtb_ubnt_e100_begin;
++extern const char __dtb_ubnt_e120_begin;
+ void __init device_tree_init(void)
+ {
+	const void *fdt;
+@@ -1206,6 +1208,14 @@ void __init device_tree_init(void)
+		fdt = &__dtb_octeon_68xx_begin;
+		do_prune = true;
+		fill_mac = true;
++	} else if (octeon_bootinfo->board_type == CVMX_BOARD_TYPE_UBNT_E100) {
```
Comment (stintel) [inline target/linux/octeon/patches-4.9/103-ubnt_usg3p_support.patch]: And this should go in the first commit, as that commit already changes things for E100. The commit that adds support for USG3P should only contains changes related to E120.

Comment (MartB) [reply target/linux/octeon/patches-4.9/103-ubnt_usg3p_support.patch]: that slipped through will get fixed.

**Comment by MartB** [comment]
Im not sure i understand the part about the patches correctly. 
None of these patches have any additional info i wrote them myself.
I thought just including the patches like the other guys did here https://github.com/lede-project/source/blob/master/target/linux/octeon/patches-4.9/100-ubnt_edgerouter2_support.patch is fine ?

**Comment by stintel** [comment]
Is this you: Written by: Aaro Koskinen <aaro.koskinen@iki.fi>?
Also, I missed it before, but your SoB line is invalid and should contain your full real name and must match the commit author. See https://lede-project.org/submitting-patches.

**Comment by MartB** [comment]
Uhm no but thats a new file with a header from the previous file i overwrote.
That needs to be deleted and replaced with the proper header i guess?

**Comment by stintel** [comment]
Yes, and you should send this upstream for review, following the guidelines for kernel contributions, and then take that patch and drop it in patches-4.9/.

**Comment by blogic** [comment]
With the remerge in progress, all PRs on the lede-project organisation will be closed. Please help getting this merged or rebase/post it on the openwrt project page (https://github.com/openwrt/openwrt/pulls). All remaining PRs will be closed in 30 days.
